File: internal/codexreport/habits.go
package codexreport

import (
	"fmt"
	"strings"
)

func buildGitHabits(agg *aggregate, repos []RepoReport) GitHabitsReport {
	var r GitHabitsReport
	for _, count := range agg.gitCommands {
		r.CommandCount += count
	}
	r.TopSubcommands = topCommands(agg.gitCommands, 10)

	branches := map[string]bool{}
	for _, repo := range repos {
		for _, branch := range repo.Branches {
			branches[repo.Repo+"@"+branch] = true
		}
		if len(repo.Branches) > 1 {
			r.MultiBranchRepos++
		}
	}
	r.BranchCount = len(branches)

	if agg.gitCommands["status"] > 0 {
		r.ReviewSignals = append(r.ReviewSignals, fmt.Sprintf("经常检查工作区状态: git status %d 次", agg.gitCommands["status"]))
	}
	if agg.gitCommands["diff"] > 0 {
		r.ReviewSignals = append(r.ReviewSignals, fmt.Sprintf("会查看差异: git diff %d 次", agg.gitCommands["diff"]))
	}
	if agg.gitCommands["log"] > 0 || agg.gitCommands["show"] > 0 {
		r.ReviewSignals = append(r.ReviewSignals, "会读取历史上下文")
	}
	if agg.gitCommands["commit"] == 0 && (agg.gitCommands["diff"] > 0 || agg.gitCommands["status"] > 0) {
		r.RiskSignals = append(r.RiskSignals, "观察到检查状态/差异，但没有提交命令；可能偏向让人类最后提交")
	}
	if agg.gitCommands["push"] > 0 {
		r.RiskSignals = append(r.RiskSignals, "观察到 push 命令；适合在 AGENTS.md 中写清推送前检查")
	}
	return r
}

func buildProjectMgmt(repos []RepoReport) ProjectMgmtReport {
	var r ProjectMgmtReport
	for _, repo := range repos {
		if len(repo.TestCommands) > 0 {
			r.ReposWithTests++
		}
		if len(repo.BuildSystems) > 0 {
			r.ReposWithBuildSystem++
		}
		if hasBuildSystem(repo.BuildSystems, "GitHub Actions") {
			r.ReposWithCI++
		}
		for _, fc := range repo.FileChangeTypes {
			switch fc.Type {
			case "Markdown":
				r.DocumentationChanges += fc.Count
			case "Project config", "YAML", "TOML", "JSON":
				r.ConfigChanges += fc.Count
			}
		}
	}
	r.PlanningFileChanges = r.DocumentationChanges

	if len(repos) > 0 {
		if r.ReposWithTests == 0 {
			r.Signals = append(r.Signals, "活跃项目中没有观察到测试命令")
		} else {
			r.Signals = append(r.Signals, fmt.Sprintf("%d/%d 个活跃项目观察到测试命令", r.ReposWithTests, len(repos)))
		}
		if r.ReposWithCI > 0 {
			r.Signals = append(r.Signals, fmt.Sprintf("%d 个活跃项目检测到 GitHub Actions", r.ReposWithCI))
		}
		if r.DocumentationChanges > 0 {
			r.Signals = append(r.Signals, fmt.Sprintf("文档/计划类变更 %d 次", r.DocumentationChanges))
		}
		if r.ConfigChanges > 0 {
			r.Signals = append(r.Signals, fmt.Sprintf("配置类变更 %d 次", r.ConfigChanges))
		}
	}
	return r
}

func hasBuildSystem(values []string, want string) bool {
	for _, value := range values {
		if strings.EqualFold(value, want) {
			return true
		}
	}
	return false
}

File: src/habits.ts
import { type CommandCount, type GitHabitsReport, type ProjectMgmtReport } from './model.js'
import { t, tf } from './i18n.js'

// 通用：把计数表转成按 count 降序、command 升序的 top-n（移植 Go topCommands 排序）。
export function topCounts(counts: Record<string, number>, n: number): CommandCount[] {
  const cc: CommandCount[] = Object.entries(counts).map(([command, count]) => ({ command, count }))
  cc.sort((a, b) => {
    if (a.count !== b.count) return b.count - a.count
    return a.command < b.command ? -1 : a.command > b.command ? 1 : 0
  })
  return cc.length > n ? cc.slice(0, n) : cc
}

export function buildGitHabits(
  gitCommands: Record<string, number>,
  branchCount: number,
  multiBranchRepos: number,
): GitHabitsReport {
  const commandCount = Object.values(gitCommands).reduce((sum, n) => sum + n, 0)
  const top = topCounts(gitCommands, 10)

  const status = gitCommands.status ?? 0
  const diff = gitCommands.diff ?? 0
  const log = gitCommands.log ?? 0
  const show = gitCommands.show ?? 0
  const commit = gitCommands.commit ?? 0
  const push = gitCommands.push ?? 0

  const reviewSignals: string[] = []
  if (status > 0) reviewSignals.push(tf('hab_review_status', { n: status }))
  if (diff > 0) reviewSignals.push(tf('hab_review_diff', { n: diff }))
  if (log > 0 || show > 0) reviewSignals.push(t('hab_review_history'))

  const riskSignals: string[] = []
  if (commit === 0 && (diff > 0 || status > 0)) {
    riskSignals.push(t('hab_risk_nocommit'))
  }
  if (push > 0) riskSignals.push(t('hab_risk_push'))

  return {
    command_count: commandCount,
    top_subcommands: top.length > 0 ? top : undefined,
    branch_count: branchCount,
    multi_branch_repos: multiBranchRepos,
    review_signals: reviewSignals.length > 0 ? reviewSignals : undefined,
    risk_signals: riskSignals.length > 0 ? riskSignals : undefined,
  }
}

export interface RepoFacts { hasTests?: boolean; hasBuild?: boolean; hasCI?: boolean }

export function buildProjectMgmt(repos: RepoFacts[]): ProjectMgmtReport {
  const reposWithTests = repos.filter(r => r.hasTests === true).length
  const reposWithBuildSystem = repos.filter(r => r.hasBuild === true).length
  const reposWithCI = repos.filter(r => r.hasCI === true).length

  const signals: string[] = []
  if (repos.length > 0) {
    if (reposWithTests === 0) signals.push(t('hab_pm_notests'))
    else signals.push(tf('hab_pm_tests', { a: reposWithTests, b: repos.length }))
    if (reposWithCI > 0) signals.push(tf('hab_pm_ci', { n: reposWithCI }))
  }

  return {
    repos_with_tests: reposWithTests,
    repos_with_build_system: reposWithBuildSystem,
    repos_with_ci: reposWithCI,
    signals: signals.length > 0 ? signals : undefined,
  }
}
